A new feature that enables the users to include headers by passing in a list

shravan97 opened the pull request:
In this PR : 
* Create a static method in `Backend` class that declares required headers
* Create a new method `include` in `__init__.py` to accept headers that need to be declared
* Declare headers in `Local` and `Dist` classes before execution
* Write tests for all above features

shravan97 committed:
Add 'declare_headers' method in Utils.py to enable users to declare C++ headers before execution

Add functionality to include required headers using ROOT Interpreter

Move ROOT import to top of the module and remove default parameter in 'declare_headers'

Change signature of 'declare_headers' to class method instead static method

Move 'declare_headers' method from 'Backend' to 'Utils'

Remove unnecessary ROOT import from Backend

Modify docstring and convert indentation to spaces in 'backend/Utils.py'

shravan97 committed:
Modify Local and Dist backend implementations to declare C++ headers before execution

Declare all headers before execution in Local and Dist backends

Import 'includes' from parent module instead of retrieving it from the Backend instance in Local and Dist

Invoke 'declare_headers' in Local and Dist from Utils

shravan97 committed:
Add a helper method 'include' in PyRDF/__init__.py to let users pass in paths to C++ headers as strings or list of strings

Add a helper method 'PyRDF.include' to enable users to include headers

Make 'includes' a module-level variable and change the signature of 'PyRDF.include'

Modify docstring in 'PyRDF.include'

shravan97 committed:
Add unit tests to check the working of 'declare_headers' method in Utils.py and 'include' method in PyRDF/__init__.py

Add tests to check the headers include feature

Modify unit tests to incorporate module changes and rename '.cpp' files to '.hxx'

Invoke 'declare_headers' from Utils instead of Backend

Clean up unit tests for 'includes' feature

Explicitly include C++ string type in 'header2.hxx'

shravan97 commented:
@etejedor and mentors, I've made `declare_headers` method static here as that cannot be serialized if it's an instance method

etejedor commented on PyRDF/backend/Backend.py:
```diff
@@ -52,6 +52,27 @@ def __init__(self, config={}):
+        self.includes = []
+
+    @staticmethod
+    def declare_headers(includes=[]):
```
Why the default to []?

shravan97 replied:
Will remove this 👍

shravan97 replied:
One more thing here @etejedor is, if `includes` is gonna be in module-level, then I don't have to pass in a parameter here and I can just import it here directly right ?

etejedor replied:
Mmm well, if this is the only place in `Backend` where you use `includes`, perhaps passing it as a parameter of `declare_headers` is fine.

shravan97 replied:
@etejedor , I can as well do an import inside the function `declare_headers` right ? `from .. import includes` ? Either way I have to do this import right ?

etejedor replied:
Or like I said, just pass it as parameter of `declare_headers` and you do not need the import.

shravan97 replied:
@etejedor , you'll need the import to get the value of `includes` right ? Inside one of the backends, before execution, in order to get the value of `includes`, you'll have to import it right ?

etejedor replied:
I do not get it now, but make the changes you find aproppriate, and I will re-review

shravan97 replied:
Ok @etejedor . Will do 👍

etejedor commented on PyRDF/backend/Backend.py:
```diff
@@ -52,6 +52,27 @@ def __init__(self, config={}):
+            to an empty list.
+
+        """
+        import ROOT
```
Should we place this import at the beginning of the module?
This helper method will be executed by the backend, which anyway will import ROOT at some point because it needs to run the computation, right? So it does not hurt to have it in the module instead of in the function.

shravan97 replied:
@etejedor . Yes, that works. Will change this 👍

etejedor commented on PyRDF/backend/Dist.py:
```diff
@@ -98,6 +98,9 @@ def execute(self, generator):
         rdf_args = generator.head_node.args
 
+        # This is to avoid serialization of 'self' object
+        includes = self.includes
```
The includes should be at main module level, not at backend level

etejedor commented on PyRDF/__init__.py:
```diff
@@ -19,4 +19,25 @@ def use(backend_name, conf = {}):
\ No newline at end of file
+        raise Exception(" Incorrect backend environment \"{}\"".format(backend))
+
+def include(includes_list=[]):
```
Proposal:
- `includes_list` is renamed to `includes` and does not have a default value  
- We have a module-level variable here called `includes` (defined right after `current_backend`) which is initialized to `[]`.
- We need to support the following sequences of operations:
  - include + use
  - use + include + use
  - include + include
- To support those sequences of operations, we need that `include` appends to the module-level list the new included headers, and that the backends read that module-level list.

shravan97 replied:
Yup @etejedor . I agree. That makes sense. I'll make these changes  👍

shravan97 replied:
Can I retain the name of `includes_list` @etejedor ? There'd be a clash in the names otherwise

etejedor replied:
Ok!

shravan97 replied:
One question @etejedor . If I declare an `includes` variable in PyRDF module level, then, how would I get the value of that variable inside a Backend class ? Is `from .. import includes` fine ?

etejedor replied:
Yes, or `from PyRDF import includes` should work too. Doesn't `Proxy` import `current_backend` already? Same thing here.

shravan97 replied:
Yup. Inside `Proxy` I'm doing a relative import. So, I have to do something similar inside `Backend` as well

etejedor commented on tests/unit/backend/test_common.py:
```diff
@@ -45,3 +45,97 @@ class TestBackend(PyRDF.backend.Backend.Backend):
+        """
+        PyRDF.current_backend = IncludeHeadersTest.TestBackend()
+
+        PyRDF.include()
```
I would not allow this case, no default value for the list

shravan97 replied:
I'll remove it @etejedor 👍

etejedor commented on tests/unit/backend/test_common.py:
```diff
@@ -45,3 +45,97 @@ class TestBackend(PyRDF.backend.Backend.Backend):
+        function when a single header needs to be included.
+
+        """
+        PyRDF.backend.Backend.Backend.declare_headers(["tests/unit/backend/test_headers/header1.cpp"])
```
Can this be `PyRDF.current_backend.declare_headers`?

etejedor replied:
We will also need to test this functionality with all the backends that we implement (Local, Spark), to make sure that the declared headers are taken when running a computation with that backend.

shravan97 replied:
`current_backend` refers to an instance of `Backend` @etejedor . But `declare_headers` is a static method. So that wouldn't work right ?

shravan97 replied:
Nop sorry. My bad. It'd still work @etejedor

shravan97 replied:
@etejedor , can I write a couple more integration tests for checking each backend after this PR ?

etejedor replied:
What if then you find a problem with your code when writing the integration tests? :) I think you can include it here, just take the time that you need.

shravan97 replied:
Ok works @etejedor . In that case, can I reuse an existing integration test or tutorial and only change one function call to a header include ?

etejedor replied:
Yes, that would work. Make a proposal and we will see :)

shravan97 replied:
Sure 👍

etejedor commented on tests/unit/backend/test_headers/header1.cpp:
```diff
@@ -0,0 +1,8 @@
+#ifndef HEADER_1
```
This and the next one should be `.hxx`

shravan97 replied:
👍

etejedor commented on tests/unit/backend/test_headers/header2.cpp:
```diff
@@ -0,0 +1,12 @@
+	return x;
+}
+
+string f2(string s) {
```
std::string?

shravan97 replied:
Yes @etejedor . ROOT takes the `std` namespace by default right ?

etejedor replied:
Yes that's correct, but it does not harm that your headers are complete (include + namespace) :D

shravan97 replied:
Sure @etejedor . Will change this 👍

shravan97 commented:
@etejedor , I have a proposal about making the `declare_headers` function non-static. If I declare a `__getstate__` (which simply returns an empty dict) and a `__setstate__` (which simply has a `pass`) inside `Backend` and make `Backend` serializable, I can use `declare_headers` as an instance method. What do you say ?

bluehood commented:
what's the problem with having a class method?

shravan97 commented:
@bluehood , not any big issue. But, I just felt it cleaner to call `self.declare_headers(...)` instead of `MyBackend.declare_headers(...)`

etejedor commented:
@shravan97 I have a couple of questions:
- Why does the `Backend` class need to be serializable?
- How does `declare_headers` affect the serializability of the `Backend` class?

etejedor commented:
@shravan97 sure, no problem. Another option would be to have `declare_headers` defined as a class method in a `Utils` class, since its functionality is not strictly related to the backend. It is actually a utility function that allows you to inject code into the interpreter.

shravan97 commented:
@etejedor , yes I agree. That's a great idea. Do you suggest that the `Utils` be inside `backend/` or  outside  ?

etejedor commented:
Let's place it inside `backend` for the moment, we can always take it out if necessary.

shravan97 commented:
Ok @etejedor . Will do that. Btw, I have to push the Spark integration test in a separate PR and here are the 2 reasons : 
* Reducer has a bug as I pointed out earlier. Python 2 has `int` and `long` types whereas Python 3 only has `int`. I've fixed this now using `numbers` module in Python though.
* There is not way to access the integer output while using Spark backend as `GetValue` would currently throw an error. There's a hacky way though : `myProxy.action_node.value`. But that doesn't sound very clean to me

shravan97 committed:
Add an integration test for Local backend to check the working of C++ headers include feature

Add a local integration test to check that headers are included correctly

Modify integration test and C++ header to retrieve all numbers less than 5

shravan97 commented:
@etejedor , made the changes now. Everything is in except an integration test for Spark. That integration test is also ready and passes. But, since it requires a change in the mapper function inside `Dist`, I'll make that a part of a separate PR. And also, the only way to access the final output value after spark execution is : `proxy.action_node.value`. Once we come up with a solution for getting the output value, I'll make the PR for that integration test

shravan97 commented:
And @etejedor , I propose these commits after rebase : 
* One for `declare_headers` function
* One for `PyRDF.include` function
* One for changes in `Local` and `Dist` classes (invoking `declare_headers`) that is.
* One for unit tests
* One for local integration test

etejedor commented on PyRDF/__init__.py:
```diff
@@ -19,4 +20,25 @@ def use(backend_name, conf = {}):
+    ----------
+    includes_list : list or str
+        This list should consist of all necessary C++
+        headers as a string. This parameter defaults
```
The parameter no longer defaults as an empty list

shravan97 replied:
Will fix this @etejedor 👍

etejedor commented on PyRDF/backend/Utils.py:
```diff
@@ -0,0 +1,24 @@
+	"""
+	@classmethod
+	def declare_headers(cls, includes):
+		"""
```
Indentation needs fix here?

etejedor commented on PyRDF/backend/Utils.py:
```diff
@@ -0,0 +1,24 @@
+            headers as a string. This parameter defaults
+            to an empty list.
+
+		"""
```
And here

etejedor commented on tests/unit/backend/test_headers/header2.hxx:
```diff
@@ -0,0 +1,12 @@
+#ifndef HEADER_2
+#define HEADER_2
+
```
`#include <string>`

shravan97 replied:
ROOT interpreter already includes `string` library right @etejedor ?

shravan97 replied:
That doesn't throw an error too

etejedor replied:
Yes, it includes it and it also makes the std namespace available. But I prefer to write complete headers that also compile without the ROOT interpreter :)

shravan97 replied:
Ok @etejedor . Will make this change 👍

etejedor commented on tests/integration/local/test_headers_include.py:
```diff
@@ -0,0 +1,28 @@
+                         .Define("b2", "(int) tdfentry_ * tdfentry_")
+
+        # This filters out all numbers less than 40 in column 'b2'
+        rdf_filtered = rdf_defined.Filter("check_number_less_than_40(b2)")
```
To make this test simpler, what about just `rdf_defined.Filter("check_number_less_than_5(tdfentry_)")`, and you do not need the previous Defines.

shravan97 replied:
Yup @etejedor . That is better. Will make this fix 👍

etejedor commented on tests/unit/backend/test_common.py:
```diff
@@ -45,3 +46,87 @@ class TestBackend(PyRDF.backend.Backend.Backend):
+
+class IncludeHeadersTest(unittest.TestCase):
+    """
+    Tests to check the working of 'PyRDF.Include' function.
```
It is `PyRDF.include`, lowercase

shravan97 replied:
👍

etejedor commented on tests/unit/backend/test_common.py:
```diff
@@ -45,3 +46,87 @@ class TestBackend(PyRDF.backend.Backend.Backend):
+
+        """
+        Utils.declare_headers(["tests/unit/backend/test_headers/header1.hxx",
+                                                       "tests/unit/backend/test_headers/header2.hxx"
```
Check indentation here

etejedor commented on PyRDF/backend/Utils.py:
```diff
@@ -0,0 +1,24 @@
+        ----------
+        includes : list
+            This list should consist of all necessary C++
+            headers as a string. This parameter defaults
```
No default anymore

etejedor commented:
Finished my review, after the last requested changes, this can go in. Ok for the structure of commits proposed.

shravan97 commented:
@etejedor , changes are done. Can I rebase now ?

etejedor commented:
Yes please

shravan97 commented:
@etejedor , rebase done and ready to be merged 👍
